### PR TITLE
1364 stock status for non msupply countries

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
[1364: 3 dashboard reports for UNFPA (Stock status for non mSupply countries)](https://github.com/beyondessential/tupaia-backlog/issues/1364)

### Changes:

- Add `tableOfCalculatedValuesForOrgUnits.js` data builder to build columns from organisation units with `tableOfCalculatedValues` data builder.
- Migration for adding three reports
- Make matching filter dynamic in `calculateOperationForAnalytics.js`
- Refactor code in `tableOfValuesForOrgUnits.js` and `tableOfDataValues.js`
- add function for row and column summary in matrix table (`addSummaryToTable.js`)

### CODE REVIEW

File: packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js [hunk 1]
```diff
@@ -24,8 +30,14 @@ const valueToGroup = (value, config) => {
...
 const performSingleAnalyticOperation = (analytics, config) => {
-  const { operator, dataElement } = config;
-  const filteredAnalytics = analytics.filter(({ dataElement: de }) => de === dataElement);
+  const { operator } = config;
+  // filterOptions could be ['dataElement', 'organisationUnit'] for multiple matching key options
+  const filterOptions = config.filterOptions ?? config.dataElement;
+  const filterOptionsValues = {};
+  filterOptions.forEach(filterOptionKey => {
+    filterOptionsValues[filterOptionKey] = config[filterOptionKey];
+  });
```
Comment (kael89) [inline packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js:39]: `Object.fromEntries` is often handy when creating value maps:

```suggestion
const valuesByFilterKey = Object.fromEntries(filterOptions.map(key => [key, config[key]]));
```

**Review by kael89 — CHANGES_REQUESTED** [review]
I have the impression that the code as is will not calculate correct percentages for `RH Products Stock Availability (HFRSA and Spot Check Survey Data) With Percentage`

File: packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/addSummaryToTable.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        },
+      },
+      excludeCondition: { value: 'No data', operator: '=' },
+    },
+ *-----------------------------------------------------
+ * 
+ */
+export const buildRowSummary = (result, config) => {
+  const newResult = result;
+  const newColumnKey = 'rowSummary';
```
Comment (kael89) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/addSummaryToTable.js:104]:
```suggestion
  const summaryColumnKey = 'rowSummary';
```

File: packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/addSummaryToTable.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    const filteredRowValuesSet = excludeCondition
+      ? rowValuesSet.filter(value => !checkValueSatisfiesCondition(value, excludeCondition))
+      : rowValuesSet;
+    newResult.rows[index][newColumnKey] = operator(filteredRowValuesSet, operatorConfig);
+  });
+
+  // Add summary column to columns object
+  const newColumn = { key: newColumnKey, title };
+  result.columns.push(newColumn);
+  return newResult;
```
Comment (kael89) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/addSummaryToTable.js:122]:
```suggestion
const newColumns = [ ...tableData.columns, { key: summaryColumnKey, title }];
return { ...tableData, rows: newRows, columns: newColumns };
```

**Review by kael89 — CHANGES_REQUESTED** [review]

**Comment by billli0** [comment]
> I have the impression that the code as is will not calculate correct percentages for `RH Products Stock Availability (HFRSA and Spot Check Survey Data) With Percentage`

Thanks Kostas! You are one of those who loves to share every knowledge. I have a big lesson from your review. 
I can tell the 'percentage' is confusing, which actually is a funciton in `addSummaryToTable.js` instead of the one in `calculateOperationForAnalytics.js`. Because this is a customer 'percentage' for the row and column. Perhaps changing the function name can make it less confusing?

**Comment by kael89** [comment]
> Thanks Kostas! You are one of those who loves to share every knowledge. I have a big lesson from your review.
> I can tell the 'percentage' is confusing, which actually is a function in `addSummaryToTable.js` instead of the one in `calculateOperationForAnalytics.js`. Because this is a customer 'percentage' for the row and column. Perhaps changing the function name can make it less confusing?

I believe it's fine @billli0 , I could tell that the `percentage` used here was the one declared in `addSummaryToTable.js`. I believe the code will not be correct because of two main issues:

* https://github.com/beyondessential/tupaia/pull/1839#discussion_r546533352 - when data exist
* https://github.com/beyondessential/tupaia/pull/1839#discussion_r546509540 - when data are absent

In fact, I wanted to be sure so I run your code locally and for FIJI, I'm getting `NaN%` for all row summaries in 2020, and `0%` for all row summaries in 2019.

**Comment by billli0** [comment]
Sorry, I miss your comments. GitHub just hides some of those for me...

File: packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/tableOfCalculatedValuesForOrgUnits.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import { TotalCalculator } from './TotalCalculator';
+import { TableOfCalculatedValuesBuilder } from './tableOfCalculatedValues';
+
+class TableOfCalculatedValuesForOrgUnitsBuilder extends TableOfCalculatedValuesBuilder {
+  async build() {
+    const { results, period } = await this.fetchAnalyticsAndMetadata();
+    this.results = results;
+    this.tableConfig = new TableConfig(this.models, this.config, this.results);
+    // Build columns first because `buildValuesByCell` function depends on it
+    const columns = await this.buildColumns();
```
[the remark this answers is not in the document]
Comment (billli0) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/tableOfCalculatedValuesForOrgUnits.js]: Thanks for your help Kostas, now I keep the chain as: 
`DataBuilder => TableOfDataValuesBuilder => TableOfCalculatedValuesBuilder`

Template design pattern is a great idea, but would like your opinion on this alternative design:

>     //buildValuesByCell() will be overided to assign value to 'this.columns'
>     this.valuesByCell = await this.buildValuesByCell(); 
>     ...
>     const columns = this.columns || (await this.buildColumns());

File: packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/helpers/addSummaryToTable.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/**
+ * Tupaia Config Server
+ * Copyright (c) 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { checkValueSatisfiesCondition } from '@tupaia/utils';
+
+const percentageOfConditions = (valueSet, config) => {
+  const valueAllUndefined = valueSet.every(value => value === undefined);
+  if (valueSet.length === 0 || valueAllUndefined) return false;
```
Comment (kael89) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/table/tableOfDataValues/helpers/addSummaryToTable.js:10]: (Detail) This check is slightly quicker: just one non `undefined` value is enough to do calculations

```suggestion
  const isAnyValueDefined = valueSet.some(value => value !== undefined);
  if (valueSet.length === 0 || !isAnyValueDefined) return false;
```

**Review by kael89 — CHANGES_REQUESTED** [review]
Looks great now @billli0 ! I prefer the pattern in your current implementation. A few changes here and there and this should be ready!

**Comment by billli0** [comment]
Thanks Kostas, I've moved `ORG_UNIT_COLUMNS_KEYS_SET` to `constant.js` to reuse it, and implemented your suggestion.

**Review by kael89 — APPROVED** [review]
Ready to go!

**Review by kael89 — APPROVED** [review]

**Review by kael89 — APPROVED** [review]

**Pull request merged**